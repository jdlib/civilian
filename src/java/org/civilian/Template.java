/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian;


import java.io.Writer;
import org.civilian.response.ResponseWriter;
import org.civilian.util.Check;


/**
 * A Template can be used to produce the textual content of a response.
 * Templates store all their dynamic data in instance variables which are initialized
 * in the template constructor:
 * Therefore a template object is self-contained and can just simply be printed 
 * to an output. This especially allows to pass templates to other templates 
 * as parameters and to build complex content by composing it from simpler parts.<br> 
 * Since template objects contain response specific data they - like Controllers - 
 * are instantiated for a single response and not reused.<br>
 * Civilian provides an own templating system and syntax (CSP), but other templates engines like
 * velocity or freemarker can easily be presented as Template objects. 
 */
public abstract class Template implements ResponseWriter.Printable
{
	/**
	 * Returns the ResponseWriter or null, if the template is not printed.
	 */
	public ResponseWriter out()
	{
		return out;
	}

	
	/**
	 * Prints the template, using the given Writer.
	 * The method constructs a ResponseWriter from the writer and
	 * then calls {@link #print(ResponseWriter)}.
	 */
	public void print(Writer out) throws Exception
	{
		Check.notNull(out, "out");
		ResponseWriter rw = out instanceof ResponseWriter ? 
			(ResponseWriter)out : 
			new ResponseWriter(out, false);
		print(rw);
	}
	
	
	/**
	 * Prints the template, using the given ResponseWriter.
	 * The method stores the ResponseWriter in the field {@link #out}
	 * and the calls {@link #print()}.
	 */
	@Override public synchronized void print(ResponseWriter out) throws Exception
	{
		Check.notNull(out, "out");
		this.out = out;
		try
		{
			print();
		}
		finally
		{
			this.out = null;
		}
	}
	
	
	/**
	 * Needs to be implemented by derived classes.
	 */
	protected abstract void print() throws Exception;
	
	
	protected ResponseWriter out;
}
