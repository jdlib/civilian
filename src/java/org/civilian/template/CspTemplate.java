/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import java.io.PrintWriter;
import org.civilian.util.Check;


/**
 * CspTemplate is a template to which uses a TemplateWriter.
 */
public abstract class CspTemplate extends Template implements TemplateWriter.Printable
{
	/**
	 * @return the TemplateWriter or null, if the template is not printed.
	 */
	public TemplateWriter out()
	{
		return out;
	}

	
	/**
	 * Prints the template, using the given Writer.
	 * The method constructs a TemplateWriter from the writer and
	 * then calls {@link #print(TemplateWriter)}.
	 * @param out a Writer
	 * @param data optioanl context data
	 * @throws Exception any exception
	 */
	@Override public void print(PrintWriter out, Object... data) throws Exception
	{
		Check.notNull(out, "out");
		if (out instanceof TemplateWriter)
			print((TemplateWriter)out);
		else
		{
			TemplateWriter tw = new TemplateWriter(out, false);
			if (data.length > 0)
				tw.getData().addAll(data);
			print(tw);
		}
	}
	
	
	/**
	 * Prints the template, using the given TemplateWriter.
	 * The method stores the TemplateWriter in the field {@link #out}
	 * and the calls {@link #print()}.
	 */
	@Override public synchronized void print(TemplateWriter out) throws Exception
	{
		Check.notNull(out, "out");
		if (this.out != null)
			throw new IllegalStateException("already printing");
		try
		{
			this.out = out;
			init();
			print();
		}
		finally
		{
			this.out = null;
			exit();
		}
	}
	
	
	/**
	 * Allows derived implementation to initialize before the template is printed .
	 * Called by {@link #print(TemplateWriter)} when {@link #out} was set, before {@link #print()} is called.
	 * The default implementation is empty.
	 */
	protected void init()
	{
	}
	
	
	/**
	 * Allows derived implementation to performe operations after the template is printed .
	 * Called by {@link #print(TemplateWriter)} after {@link #print()} was called.
	 * The default implementation is empty.
	 */
	protected void exit()
	{
	}

	
	/**
	 * Needs to be implemented by derived classes.
	 * @throws Exception any exception
	 */
	protected abstract void print() throws Exception;
	
	
	protected TemplateWriter out;
}
