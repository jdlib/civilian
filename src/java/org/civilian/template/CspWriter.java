/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import java.io.IOException;
import java.io.PrintWriter;
import java.io.Writer;
import org.civilian.util.Check;
import org.civilian.util.ClosedWriter;
import org.civilian.util.Data;


/**
 * CspWriter is a PrintWriter to write pretty indented files.
 * CspWriter maintains a {@link #getTabCount() tab count}. When a new line is started
 * then tab characters are automatically inserted at the beginning of the line, according
 * to the tab count. The default tab characters is a single '\t' character, but you can
 * chose any string instead (e.g. "  "). 
 */
public class CspWriter extends PrintWriter
{
	/**
	 * Sets the default characters used by a new CspWriter to indent a line.
	 * By default a indent consists of a single tab character.
	 * @param chars the chars
	 * @see #setTabChars(String)
	 */ 
	public static void setDefaultTabChars(String chars)
	{
		defaultTabChars_ = getChars(chars, "chars");
	}

	
	/**
	 * @return the default characters used by a new CspWriter to indent a line.
	 */ 
	public static String getDefaultTabChars()
	{
		return new String(defaultTabChars_);
	}

	
	/**
	 * Sets the default line separator used by a new CspWriter.
	 * By default this is '\n' (in accordance to our primary goal to produce web content.
	 * @param separator the separator
	 */ 
	public static void setDefaultLineSeparator(String separator)
	{
		defaultLineSeparator_ = getChars(separator, "separator");
	}

	
	/**
	 * Returns the default line separator used by a new CspWriter.
	 */ 
	public static String getDefaultLineSeparator()
	{
		return new String(defaultLineSeparator_);
	}

	
	/**
	 * Creates a new CspWriter.
	 * @param out a Writer
	 */
	public CspWriter(Writer out)
	{
		this(out, false);
	}


	/**
	 * Creates a CspWriter.
	 * @param out a Writer
	 * @param autoFlush - a boolean; if true, the println() methods will flush
	 *      the output buffer
	 */
	public CspWriter(Writer out, boolean autoFlush)
	{
		super(out, false);
		autoFlush_ = autoFlush;
	}


	/**
	 * Sets the string used for a tab. 
	 * If you do not explicitly specify the tab chars, 
	 * the value of {@link #getDefaultTabChars()} is used.
	 */
	public void setTabChars(String chars)
	{
		tabChars_ = getChars(chars, "chars");
	}


	/**
	 * Sets the characters used to separate lines. The default is
	 * system dependent (e.g. "0xD0xA" on Windows).
	 */
	public void setLineSeparator(String separator)
	{
		lineSeparator_ = getChars(separator, "separator");
	}


	/**
	 * Increases the indent tab.
	 */
	public void increaseTab()
	{
		tabCount_++;
	}


	/**
	 * Decreases the indent tab.
	 */
	public void decreaseTab()
	{
		if (tabCount_ > 0)
			tabCount_--;
	}


	/**
	 * @return the current number of indent tabs.
	 */
	public int getTabCount()
	{
		return tabCount_;
	}


	public void setTabCount(int count)
	{
		if (count < 0)
			throw new IllegalArgumentException();
		tabCount_ = count;
	}

	
	protected void writeNewLineTab()
	{
		for (int j=tabCount_; j>0; j--)
			super.write(tabChars_, 0, tabChars_.length);
		newLineStarted_ = false;
	}


	public boolean newLineStarted()
	{
		return newLineStarted_;
	}


	//-------------------------------------------------------
	// In fact the whole printwriter class is duplicated here
	// since this class should be a printwriter but
	// not increase the performance overhead significantly
	// because of the newline mechanism
	//-------------------------------------------------------
	

	/**
	 * Closes the stream.
	 */
	@Override public void close()
	{
		try
		{
			flush();
			synchronized(lock)
			{
				Writer out = this.out;
				this.out   = ClosedWriter.INSTANCE;
				if (out != null)
					out.close();
			}
		}
		catch(IOException e)
		{
			setError(e);
		}
	}


	/**
	 * Flushes the stream and check its error state.  Errors are cumulative;
	 * once the stream encounters an error, this routine will return true on
	 * all successive calls.
	 * @return true if the print stream has encountered an error, either on the
	 * underlying output stream or during a format conversion.
	 */
	@Override public boolean checkError()
	{
		if (out != ClosedWriter.INSTANCE)
			flush();
		return error_ != null;
	}


	/**
	 * Indicates that an error has occurred.
	 * @param e the error
	 */
	protected void setError(IOException e)
	{
		if (error_ == null)
			error_ = e;
	}

	
	/**
	 * @return the error or null if none has occurred.
	 */
	public IOException getError()
	{
		return error_;
	}
	

	/**
	 * Writes a single character.
	 * @param c the character
	 */
	@Override public void write(int c)
	{
		if (newLineStarted_)
			writeNewLineTab();
		super.write(c);
	}


	/**
	 * Writes a portion of an array of characters.
	 * @param buf a buffer
	 * @param off the offset
	 * @param length the length to write
	 */
	@Override public void write(char buf[], int off, int len)
	{
		if (newLineStarted_)
			writeNewLineTab();
		super.write(buf, off, len);
	}


	/**
	 * Writes a portion of a string.
	 * @param s a String
	 * @param off the offset
	 * @param len the length to write
	 */
	@Override public void write(String s, int off, int len)
	{
		if (newLineStarted_)
			writeNewLineTab();
		super.write(s, off, len);
	}


	/**
	 * Writes a string. This method cannot be inherited from the Writer class
	 * because it must suppress I/O exceptions.
	 * @param s a String
	 */
	@Override public void write(String s)
	{
		write(s, 0, s.length());
	}


	@Override public void print(char c)
	{
		write(c);
	}


	/**
	 * Write a line end, but only if the current line is not empty.
	 */
	public void printlnIfNotEmpty()
	{
		if (!newLineStarted_)
			println();
	}
	
	
	@Override public void println()
	{
		// empty lines are not indented
		if (newLineStarted_)
			newLineStarted_ = false;
		write(lineSeparator_, 0, lineSeparator_.length);
		if (autoFlush_)
			flush();
		newLineStarted_ = true;
	}


	@Override public void println(boolean value)
	{
		print(value);
		println();
	}


	@Override public void println(char value)
	{
		print(value);
		println();
	}


	@Override public void println(int value)
	{
		print(value);
		println();
	}


	@Override public void println(long value)
	{
		print(value);
		println();
	}


	@Override public void println(float value)
	{
		print(value);
		println();
	}


	@Override public void println(double value)
	{
		print(value);
		println();
	}


	@Override public void println(char[] value)
	{
		print(value);
		println();
	}


	@Override public void println(String value)
	{
		print(value);
		println();
	}


	@Override public void println(Object value)
	{
		print(value);
		println();
	}


	/**
	 * Checks if the object is a printable and in that case calls
	 * print(Printable), else it just calls the default implementation.
	 * @param object an object
	 */
	@Override public void print(Object object)
	{
		if (object instanceof Printable)
			print((Printable)object);
		else
			super.print(object);
	}
	
	
	/**
	 * If a not-null Printable is passed to the CspWriter,
	 * then the printable is asked to print itself. 
	 * @param printable a printable
	 */
	public void print(Printable printable)
	{
		if (printable != null)
		{
			try
			{
				printable.print(this);
			}
			catch (RuntimeException e)
			{
				throw e;
			}
			catch (Exception e)
			{
				throw new IllegalStateException("error when printing '" + printable.getClass().getName() + "'", e);
			}
		}
	}
	
	
	/**
	 * Printable is a interface for print-aware classes who 
	 * implement a custom print strategy. Templates and form controls
	 * are examples of Printables.
	 * If you pass a Printable to {@link CspWriter#print(Object)} or
	 * CspWriter#print(Printable), then the {@link #print(CspWriter)}
	 * method of the Printable is called by the CspWriter, allowing
	 * to Printable to print itself.
	 */
	public static interface Printable 
	{
		public void print(CspWriter out) throws Exception;
	}

	
	private static char[] getChars(String s, String what)
	{
		Check.notNull(s, what);
		return s.toCharArray();
	}
	
	
	public Data getData()
	{
		if (data_ == null)
			data_ = new Data();
		return data_;
	}
	
	
	/**
	 * @return toString() of the wrapped writer. 
	 */
	@Override public String toString()
	{
		return out.toString();
	}


	private boolean newLineStarted_ = true;
	private int tabCount_ = 0;
	private char tabChars_[] = defaultTabChars_;
	private char lineSeparator_[] = defaultLineSeparator_;
	private boolean autoFlush_;
	private IOException error_;
	private Data data_;
	private static char[] defaultTabChars_ = { '\t' };
	private static char[] defaultLineSeparator_ = { '\n' };
}
