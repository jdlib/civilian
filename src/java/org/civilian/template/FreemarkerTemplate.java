/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import java.io.IOException;


/**
 * A Template implementation which is based on a Freemarker template.
 */
public class FreemarkerTemplate extends org.civilian.template.Template
{
	/**
	 * Creates a new FreemarkerTemplate. 	 
	 * @param template the freemarker tempate
	 * @param model the model
	 */
	public FreemarkerTemplate(freemarker.template.Template template, Object model)
	{
		template_ 	= template;
		model_ 		= model;
	}
	
	
	@Override protected void print() throws IOException, freemarker.template.TemplateException
	{
		template_.process(model_, out);  
	}

	
	private Object model_;
	private freemarker.template.Template template_;
}
