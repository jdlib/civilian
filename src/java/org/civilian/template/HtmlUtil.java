/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import java.io.PrintWriter;
import org.civilian.util.Charset;
import org.civilian.util.StringUtil;


/**
 * Provides utility methods to print HTML attributes and text
 */
public class HtmlUtil
{
	/**
	 * Prints a String attribute.
	 * The output is: ' ' name '="' escaped-value '"'.
	 * @param out a PrintWriter
	 * @param name the attribute name
	 * @param value the attribute value
	 */
	public static void attr(PrintWriter out,String name, String value)
	{
		attr(out, name, value, true);
	}


	/**
	 * Prints an attribute.
	 * The output is: ' ' name '="'value '"'.
	 * @param out a PrintWriter
	 * @param name the attribute name
	 * @param value the attribute value
	 * @param escape should the attribute value be escaped. Only pass false, if you
	 * 		know that escaping is not needed.
	 */
	public static void attr(PrintWriter out, String name, String value, boolean escape)
	{
		if (name != null)
		{
			out.print(' ');
			out.print(name);
			out.print("=\"");
			if (value != null)
			{
				if (escape)
					escape(out, value, true);
				else
					out.print(value);
			}
			out.print('"');
		}
	}
	
	
	/**
	 * Prints a integer attribute.
	 * The output is: ' ' name '="' value '"'.
	 * @param out a PrintWriter
	 * @param name the attribute name
	 * @param value the attribute value
	 */
	public static void attr(PrintWriter out,String name, int value)
	{
		attr(out, name, String.valueOf(value), false);
	}
	
	
	/**
	 * Prints attributes.
	 * @param out a PrintWriter
	 * @param attrs a list of attribute pairs, i.e. a name string immediately followed by its value string.
	 */
	public static void attrs(PrintWriter out, String... attrs)
	{
		if (attrs != null)
		{
			for (int i=0; i<attrs.length; )
			{
				String name 	= attrs[i++];
				String value 	= i < attrs.length ? attrs[i++] : ""; 
				attr(out, name, value);
			}
		}
	}


	/**
	 * Escapes and prints the text.
	 * @param out a PrintWriter
	 * @param text the text
	 */
	public static void text(PrintWriter out, String text)
	{
		escape(out, text, false);
	}


	/**
	 * Escapes a string.
	 * @param out a PrintWriter
	 * @param input the string
	 * @param isAttribute true, if escape rules for attribute, false if for text values should be applied
	 */
	public static void escape(PrintWriter out, String input, boolean isAttribute)
	{
		escape(out, input, isAttribute, null);
	}
	
	
	/**
	 * Escapes a string.
	 * @param out receives the escaped string
	 * @param input the string
	 * @param isAttribute true, if escape rules for attribute, false if for text values should be applied
	 * @param charset if not null, the charset is used to determine if a character is printable
	 * 		in the current encoding. If not a HTML character is printed instead. 
	 */
	public static void escape(PrintWriter out, String input, boolean isAttribute, Charset charset)
	{
		if (input == null)
			return;
		
		int length = input.length();

		for (int i=0; i<length; i++)
		{
			char c = input.charAt(i);
			switch(c)
			{
				case '"': out.print("&quot;");	break;
				case '&': out.print("&amp;");	break;
				case '<': out.print("&lt;");	break;
				case '>': out.print("&gt;");	break;
				case '\n':
					if (isAttribute)
						out.print("&#10;");
					else
						out.print(c);
					break;
				case '\r':
					if (!isAttribute)
						out.print(c);
					continue;
				default:
					if ((charset != null) && !charset.isPrintable(c))
						out.print("&#" + String.valueOf((int)c) + ';');
					else
						out.print(c);
					break;
			}
		}
	}


	/**
	 * Prints a JavaScript string which is embedded in a HTML page.
	 * It escapes \', \n, \r \t and \\ characters and converts
	 * character which are not printable in the current encoding to
	 * a HTML character reference. It prints null, if the text is null
	 * @param out a PrintWriter
	 * @param text the string
	 * @param addQuotes adds single quote character around the string if true. 
	 */
	public static void jsString(PrintWriter out, String text, boolean addQuotes)
	{
		if (text == null)
			out.print("null");
		else
		{
			if (addQuotes)
				out.print("'");
		
			int length = text.length();
			int start  = 0;
			boolean printCharRef = false;
			String ref = null;
	
			for (int i=0; i<length; i++)
			{
				char c = text.charAt(i);
				switch(c)
				{
					case '\'': ref = "\\\'";	break;
					case '\n': ref = "\\n";		break;
					case '\r': ref = "\\r";		break;
					case '\t': ref = "\\t";		break;
					case '\\': ref = "\\\\";	break;
					default:
						continue;
				}
				out.write(text, start, i - start);
				if (printCharRef)
				{
					out.print("\\u");
					out.print(StringUtil.fillLeft(c, 4));
					out.print(';');
					printCharRef = false;
				}
				else
					out.print(ref);
				start = i+1;
			}	
			out.write(text, start, length - start);
			if (addQuotes)
				out.print("'");
		}
	}
}
