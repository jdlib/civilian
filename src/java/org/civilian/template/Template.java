/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import java.io.StringWriter;
import java.io.Writer;
import org.civilian.util.Check;


/**
 * A Template can be used to produce the textual content of a response.
 * Templates store all their dynamic data in instance variables which are initialized
 * in the template constructor:
 * Therefore a template object is self-contained and can just simply be printed 
 * to an output. This especially allows to pass templates to other templates 
 * as parameters and to build complex content by composing it from simpler parts.<br> 
 * Since template objects contain response specific data they - like Controllers - 
 * are instantiated for a single response and not reused.<br>
 * Civilian provides an own templating system and syntax (CSP), but other templates engines like
 * velocity or freemarker can easily be presented as Template objects. 
 */
public abstract class Template implements TemplateWriter.Printable
{
	/**
	 * @return the TemplateWriter or null, if the template is not printed.
	 */
	public TemplateWriter out()
	{
		return out;
	}

	
	/**
	 * Prints the template and returns it as string.
	 * @return the string
	 * @throws Exception any exception allowed
	 */
	public String printString() throws Exception
	{
		StringWriter out = new StringWriter();
		print(out);
		return out.toString();
	}
	
	
	/**
	 * Prints the template, using the given Writer.
	 * The method constructs a TemplateWriter from the writer and
	 * then calls {@link #print(TemplateWriter)}.
	 * @param out a Writer
	 * @throws Exception any exception
	 */
	public void print(Writer out) throws Exception
	{
		Check.notNull(out, "out");
		TemplateWriter tw = out instanceof TemplateWriter ? 
			(TemplateWriter)out : 
			new TemplateWriter(out, false);
		print(tw);
	}
	
	
	/**
	 * Prints the template, using the given TemplateWriter.
	 * The method stores the TemplateWriter in the field {@link #out}
	 * and the calls {@link #print()}.
	 */
	@Override public synchronized void print(TemplateWriter out) throws Exception
	{
		Check.notNull(out, "out");
		if (this.out != null)
			throw new IllegalStateException("already printing");
		try
		{
			this.out = out;
			init();
			print();
		}
		finally
		{
			this.out = null;
			exit();
		}
	}
	
	
	/**
	 * Allows derived implementation to initialize before the template is printed .
	 * Called by {@link #print(TemplateWriter)} when {@link #out} was set, before {@link #print()} is called.
	 * The default implementation is empty.
	 */
	protected void init()
	{
	}
	
	
	/**
	 * Allows derived implementation to performe operations after the template is printed .
	 * Called by {@link #print(TemplateWriter)} after {@link #print()} was called.
	 * The default implementation is empty.
	 */
	protected void exit()
	{
	}

	
	/**
	 * Needs to be implemented by derived classes.
	 * @throws Exception any exception
	 */
	protected abstract void print() throws Exception;
	
	
	protected TemplateWriter out;
}
