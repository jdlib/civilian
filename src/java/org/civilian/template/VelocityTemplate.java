/*
 * Copyright (C) 2014 Civilian Framework.
 *
 * Licensed under the Civilian License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.civilian-framework.org/license.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.civilian.template;


import org.apache.velocity.VelocityContext;
import org.apache.velocity.app.Velocity;
import org.civilian.util.Check;


/**
 * A Template implementation which uses a Velocity template.
 */
public class VelocityTemplate extends org.civilian.template.Template
{
	public VelocityTemplate(String templateFile, VelocityContext context)
	{
		this(Velocity.getTemplate(templateFile), context);
	}
	
	
	public VelocityTemplate(org.apache.velocity.Template template, VelocityContext context)
	{	
		template_	= Check.notNull(template, "template");
		context_ 	= context;
	}

	
	@Override protected void print() throws Exception
	{
		template_.merge(context_, out);
	}
	
	
	private final org.apache.velocity.Template template_;
	private final VelocityContext context_;
}
