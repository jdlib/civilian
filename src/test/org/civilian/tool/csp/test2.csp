java
package org.civilian.template.csp;


import java.util.ArrayList;
import org.civilian.template.Template;


public class Test2 extends Template
{
	public Test2(String name)
	{
		this.name = name;
	}
	
	
	@Override protected void print() throws Exception
	{{
		<start>
		@for (int i=0; i<10; i++)
			^?{i % 2 == 0}|conditional|
			^name: ^i 	
			@list_.add(name);
		</start>		
	}}


	private String name; 
	private ArrayList<String> list_ = new ArrayList<>();
}