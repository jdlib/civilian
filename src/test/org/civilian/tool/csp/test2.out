package org.civilian.template.csp;


import java.util.ArrayList;
import org.civilian.template.Template;


public class Test2 extends Template
{
	public Test2(String name)
	{
		this.name = name;
	}


	@Override protected void print() throws Exception
	{
		out.println("<start>");                                         // line 19: <start>
		for (int i=0; i<10; i++)                                        // line 20: @for (int i=0; i<10; i++)
		{
			if (i % 2 == 0)                                             // line 21: ^?{i % 2 == 0}
			{
				out.print("conditional");                               // line 21: conditional
			}
			out.printlnIfNotEmpty();
			out.print(name);                                            // line 22: ^name
			out.print(": ");                                            // line 22: :
			out.print(i);                                               // line 22: ^i
			out.printlnIfNotEmpty();
			list_.add(name);                                            // line 23: @list_.add(name);
		}
		out.println("</start>");                                        // line 24: </start>
	}


	private String name;
	private ArrayList<String> list_ = new ArrayList<>();
}
